qiskit/providers/aer/library/save_expectation_value.py

```diff
@@ -0,0 +1,246 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Simulator instruction to save exact operator expectation value.
+"""
+
+from numpy import allclose
+from qiskit.quantum_info import Pauli, SparsePauliOp, Operator
+from qiskit.circuit import QuantumCircuit
+from qiskit.extensions.exceptions import ExtensionError
+from .save_data import SaveAverageData
+
+
+class SaveExpectationValue(SaveAverageData):
+    """Save expectation value of an operator."""
+    def __init__(self,
+                 key,
+                 operator,
+                 unnormalized=False,
+                 conditional=False,
+                 pershot=False):
+        r"""Instruction to save the expectation value of a Hermitian operator.
+
+        The expectation value of a Hermitian operator :math:`H` for a simulator
+        in quantum state :math`\rho`is given by
+        :math:`\langle H\rangle = \mbox{Tr}[H.\rho]`.
+
+        Args:
+            key (str): the key for retrieving saved data from results.
+            operator (Pauli or SparsePauliOp or Operator): a Hermitian operator.
+            unnormalized (bool): If True return save the unnormalized accumulated
+                                 or conditional accumulated expectation value
+                                 over all shot [Default: False].
+            pershot (bool): if True save a list of expectation values for each shot
+                            of the simulation rather than the average over
+                            all shots [Default: False].
+            conditional (bool): if True save the average or pershot data
+                                conditional on the current classical register
+                                values [Default: False].
+
+        Raises:
+            ExtensionError: if the input operator is invalid or not Hermitian.
+
+        .. note:
+
+            This instruction can be directly appended to a circuit using the
+            :func:`save_expectation_value` circuit method.
+        """
+        # Convert O to SparsePauliOp representation
+        if isinstance(operator, Pauli):
+            operator = SparsePauliOp(operator)
+        elif not isinstance(operator, SparsePauliOp):
+            operator = SparsePauliOp.from_operator(Operator(operator))
+        if not isinstance(operator, SparsePauliOp):
+            raise ExtensionError("Invalid input operator")
```

Comment:
Can we reach this code? Aren't we always transforming to a SparePauliOp?

Reply:
I think you're right